arch/xtensa/src/esp32s3/esp32s3_psram.c

```diff
@@ -0,0 +1,546 @@
+{
+  static uint32_t current_wrap_size = 0;
+
+  if (current_wrap_size == wrap_size)
+    {
+      return OK;
+    }
+
+  switch (wrap_size)
+    {
+      case 32:
+      case 0:
+          psram_set_wrap_burst_length(1, PSRAM_CMD_QPI);
+          current_wrap_size = wrap_size;
+          return OK;
+      case 16:
+      case 64:
+      default:
+          return -EFAULT;
+    }
+}
+
+bool psram_support_wrap_size(uint32_t wrap_size)
+{
+  switch (wrap_size)
+    {
+      case 0:
+      case 32:
+          return true;
+      case 16:
+      case 64:
+      default:
+          return false;
+    }
+}
+
+/* Read ID operation only supports SPI CMD and mode, should issue
+ * `psram_disable_qio_mode` before calling this
+ */
+
+static void psram_read_id(int spi_num, uint32_t *dev_id)
+{
+  psram_exec_cmd(spi_num, PSRAM_CMD_SPI,
+  PSRAM_DEVICE_ID, 8,               /* command and command bit len */
+  0, 24,                            /* address and address bit len */
+  0,                                /* dummy bit len */
+  NULL, 0,                          /* tx data and tx bit len */
+  (uint8_t *) dev_id, 24,           /* rx data and rx bit len */
+  CS_PSRAM_SEL,                     /* cs bit mask */
+  false);                           /* whether is program/erase operation */
+}
+
+/* enter QPI mode */
+
+static void psram_enable_qio_mode(int spi_num)
+{
+  psram_exec_cmd(spi_num, PSRAM_CMD_SPI,
+  PSRAM_ENTER_QMODE, 8,             /* command and command bit len */
+  0, 0,                             /* address and address bit len */
+  0,                                /* dummy bit len */
+  NULL, 0,                          /* tx data and tx bit len */
+  NULL, 0,                          /* rx data and rx bit len */
+  CS_PSRAM_SEL,                     /* cs bit mask */
+  false);                           /* whether is program/erase operation */
+}
+
+static void psram_set_cs_timing(void)
+{
+  /* SPI0/1 share the cs_hold / cs_setup, cd_hold_time / cd_setup_time
+   * registers for PSRAM, so we only need to set SPI0 related registers here
+   */
+
+  SET_PERI_REG_BITS(SPI_MEM_SPI_SMEM_AC_REG(0),
+                    SPI_MEM_SPI_SMEM_CS_HOLD_TIME_V, 0,
+                    SPI_MEM_SPI_SMEM_CS_HOLD_TIME_S);
+  SET_PERI_REG_BITS(SPI_MEM_SPI_SMEM_AC_REG(0),
+                    SPI_MEM_SPI_SMEM_CS_SETUP_TIME_V, 0,
+                    SPI_MEM_SPI_SMEM_CS_SETUP_TIME_S);
+  SET_PERI_REG_MASK(SPI_MEM_SPI_SMEM_AC_REG(0),
+                    SPI_MEM_SPI_SMEM_CS_HOLD_M | \
+                    SPI_MEM_SPI_SMEM_CS_SETUP_M);
+}
+
+static void psram_gpio_config(void)
+{
+  /* CS1 */
+
+  uint8_t cs1_io = PSRAM_CS_IO;
+
+  if (cs1_io == SPI_CS1_GPIO_NUM)
+    {
+      esp32s3_gpio_matrix_out(cs1_io, SPICS1_OUT_IDX, 0, 0);
+    }
+  else
+    {
+      esp32s3_configgpio(cs1_io, OUTPUT);
+      esp32s3_gpio_matrix_out(cs1_io, SIG_GPIO_OUT_IDX, 0, 0);
+    }
+
+  g_psram_cs_io = cs1_io;
+
+  /* WP HD */
+
+  uint8_t wp_io = PSRAM_SPIWP_SD3_IO;
```

Comment:
can we move declaration to the beginning of the function?

Reply:
We can, but arch/ code don't need to follow the C89 because normally a more recent compiler is available for these chips

Reply:
This is just a suggestion. I'm fine with existing code